crates/pera-core/src/signature_mpc/mpc_manager.rs

```diff
@@ -0,0 +1,450 @@
+    type InitEvent: MPCEvent + Serialize + for<'a> Deserialize<'a>;
+    /// The MPC Manager will finalize the mpc instance after the finalize event is received.
+    type FinalizeEvent: MPCEvent + Serialize + for<'a> Deserialize<'a>;
+
+    fn new(parties: HashSet<PartyID>, party_id: PartyID) -> Self;
+}
+
+/// Convert a given authority name (address) to it's corresponding party ID.
+/// The party ID is the index of the authority in the committee.
+fn authority_name_to_party_id(
+    authority_name: AuthorityName,
+    epoch_store: &AuthorityPerEpochStore,
+) -> PeraResult<PartyID> {
+    Ok(epoch_store
+        .committee()
+        .authority_index(&authority_name)
+        // This should never happen, as the validator only accepts messages from committee members
+        .ok_or_else(|| {
+            PeraError::InvalidCommittee(
+                "Received a proof MPC message from a validator that is not in the committee"
+                    .to_string(),
+            )
+        })? as PartyID)
+}
+
+/// A Signature MPC session instance
+/// It keeps track of the status of the session, the channel to send messages to the instance,
+/// and the messages that are pending to be sent to the instance.
+struct SignatureMPCInstance<P: CreatableParty> {
+    status: MPCSessionStatus<P::OutputValue>,
+    /// The messages that are pending to be executed while advancing the instance
+    /// We need to accumulate threshold of those before advancing the instance
+    pending_messages: HashMap<PartyID, P::Message>,
+    consensus_adapter: Arc<dyn SubmitToConsensus>,
+    epoch_store: Weak<AuthorityPerEpochStore>,
+    epoch_id: EpochId,
+    /// The total number of parties in the chain
+    /// We can calculate the threshold and parties IDs (indexes) from it
+    /// To calculate the parties IDs all we need to know is the number of parties, as the IDs are just the indexes of those parties. If there are 3 parties, the IDs are [0, 1, 2].
+    number_of_parties: usize,
+    session_id: ObjectID,
+    sender_address: PeraAddress,
+    /// The MPC party that being used to run the MPC cryptographic steps. An option because it can be None before the instance has started.
+    party: Option<P>,
+}
+
+impl<P: CreatableParty> SignatureMPCInstance<P> {
+    fn new(
+        consensus_adapter: Arc<dyn SubmitToConsensus>,
+        epoch_store: Weak<AuthorityPerEpochStore>,
+        epoch: EpochId,
+        session_id: ObjectID,
+        sender_address: PeraAddress,
+        number_of_parties: usize,
+    ) -> Self {
+        Self {
+            status: MPCSessionStatus::Active,
+            pending_messages: HashMap::new(),
+            consensus_adapter: consensus_adapter.clone(),
+            epoch_store: epoch_store.clone(),
+            epoch_id: epoch,
+            session_id,
+            sender_address,
+            party: None,
+            number_of_parties,
+        }
+    }
```

Comment:
```suggestion
```
I disagree with the design here. There are two parts that aren't generic:
- party instantiation
- auxiliary input creation
These are flow-dependent, and shouldn't be handled here generically; instead, they should be handled in the the flow that the transaction that requested the MPC session instantiated. Then, you should pass the auxiliary input + party to the manager (?).

Reply:
As we agreed we'll change this in a later PR

Reply:
It's already in the code for the DKG PR, we won't add an issue this time since it's implemented already.
The PR will be published later this week.